docs/attributes-registry/code.md

```diff
@@ -6,15 +6,28 @@
 
 # Code
 
+- [Code Attributes](#code-attributes)
+- [Deprecated Code Attributes](#deprecated-code-attributes)
+
 ## Code Attributes
 
-These attributes allow to report this unit of code and therefore to provide more context about the span.
+These attributes provide context about source code
 
 | Attribute | Type | Description | Examples | Stability |
 |---|---|---|---|---|
-| <a id="code-column" href="#code-column">`code.column`</a> | int | The column number in `code.filepath` best representing the operation. It SHOULD point within the code unit named in `code.function`. | `16` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+| <a id="code-column-number" href="#code-column-number">`code.column.number`</a> | int | The column number in `code.filepath` best representing the operation. It SHOULD point within the code unit named in `code.function`. | `16` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
 | <a id="code-filepath" href="#code-filepath">`code.filepath`</a> | string | The source code file name that identifies the code unit as uniquely as possible (preferably an absolute file path). | `/usr/local/MyApplication/content_root/app/index.php` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
-| <a id="code-function" href="#code-function">`code.function`</a> | string | The method or function name, or equivalent (usually rightmost part of the code unit's name). | `serveRequest` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
-| <a id="code-lineno" href="#code-lineno">`code.lineno`</a> | int | The line number in `code.filepath` best representing the operation. It SHOULD point within the code unit named in `code.function`. | `42` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
-| <a id="code-namespace" href="#code-namespace">`code.namespace`</a> | string | The "namespace" within which `code.function` is defined. Usually the qualified class or module name, such that `code.namespace` + some separator + `code.function` form a unique identifier for the code unit. | `com.example.MyHttpService` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+| <a id="code-function-name" href="#code-function-name">`code.function.name`</a> | string | The method or function name, or equivalent (usually rightmost part of the code unit's name). | `serveRequest` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```

Comment:
The description says that it can be also something "equivalent".

Therefore, [`otelslog`](https://pkg.go.dev/go.opentelemetry.io/contrib/bridges/otelslog) and [`otelzap`](https://pkg.go.dev/go.opentelemetry.io/contrib/bridges/otelzap) logging bridges set the value to a fully-qualified name e.g. `go.opentelemetry.io/contrib/bridges/otelslog.TestSLogHandler`. Given the attribute was also called `code.function` if felt pretty acceptable and compliant (maybe we were wrong).

We could set `code.function` to `TestSLogHandler` and add `code.namespace` with `go.opentelemetry.io/contrib/bridges/otelslog`. However, it would increase the performance overhead (as the Go APIs return and operate on fully-qualified names) and felt unnecessary. Should we split it anyway? Should there be a separate attribute for a fully qualified name? Or is it OK to use `code.function.name` for fully-qualified names?

I think we should have a clarification for such use case if we are planning to make breaking changes in these attributes.

Reply:
CC @Jesse0Michael @suniastar @dmathieu @MrAlias @scorpionknifes

Reply:
We should probably better clarify with examples for each platform/language what are the expected values for `code.namespace`, `code.function.name`. Leaving this for interpretation is not a great way to ensure consistency in the future. The downside of doing this though is that we will need input from lots of people.

I _think_ the original intent (this was contributed from ECS) is being able to search/group things by attributes, so it means that having things split by default is easier. Without this, it means having to do a prefix/suffix search and grouping which might not be available for all backends, also it means having some per-platform/language knowledge on the structure of a string attribute, in other words it's stringly-typed.

What orders of performance overhead do you expect to see here ? I'm not familiar with Go, but in Java it would mostly be splitting a string.

Reply:
this reminds me of our choice around capturing `url.full` on http client (https://github.com/open-telemetry/opentelemetry-specification/pull/2469). we decided it was very helpful for consumers of the data to have consistency instead of having some http clients reporting `url.full` and some reporting `url.scheme`, `url.path`, `url.query`, even though this can incur overhead in http client instrumentations that don't have direct access to `url.full` and have to construct it from `url.scheme`, `url.path`, `url.query`.

I think here it would be better for consumers of the data if we capture `code.namespace` and `code.function` consistently (either always splitting them or always combining them).

are folks aware of other instrumentation besides the Go logging bridges that capture `code.function` in this combined form?

if we find it's a limited set of instrumentation, and we decide to recommend splitting, there are still a couple of options if the Go logging bridges can't accept the splitting overhead:

* document in the Go logging bridges that they don't follow the `code.*` semantic conventions and that for performance reasons they report the combined value in `code.function.name` (or `code.namespace`?), and maybe give users an option if they prefer for the logging bridge to do the splitting
* add a new `code.*` semantic attribute that represents the combined value, though my initial thought is this may be less desirable for similar reasons to [open-telemetry/opentelemetry-specification#2469](https://github.com/open-telemetry/opentelemetry-specification/pull/2469) in that it multiplies the combinations backends need to deal with

Reply:
I will do some benchmarks for Go. The overhead may be acceptable for us especially that it is an opt-in feature) but I wanted to call it out as the same concern could be valid for other languages.

Reply:
> if we find it's a limited set of instrumentation,

It might be applicable for all Go instrumentations and logging bridges.

For Go we can simply decide to not set `code.namespace` and set the `code.function.name` with the "full function name" with the argument that this is how the Go runtime returns the function name:
- https://pkg.go.dev/runtime#Func.Name
- https://go.dev/play/p/As0a5RYy0sJ

This is also how existing Go logging libraries (e.g. https://pkg.go.dev/go.uber.org/zap and https://pkg.go.dev/log/slog from the standard library) report the function names.

I think this is how many Go end-users would like the logging bridges to report the "function names".

However, I am worried that setting it as  `code.function.name` may be confusing.

What is more to add, the [Go Specification](https://go.dev/ref/spec#Function_declarations) also defines a function name which is what you would expect from a function name.

I want also to call out the doc comment from https://pkg.go.dev/log/slog#Source:

```go
type Source struct {
	// Function is the package path-qualified function name containing the
	// source line. If non-empty, this string uniquely identifies a single
	// function in the program. This may be the empty string if not known.
	Function [string](https://pkg.go.dev/builtin#string) `json:"function"`
```



> add a new code.* semantic attribute that represents the combined value

This is what I would propose. Notice that this is just an "attribute registry" document and not actual instrumentation library semantic convention. The logging bridges may be configurable whether they emit `code.function.name` + `code.registry` or `code.function.full-name`.

Reply:
Thanks for doing go benchmarks @pellared, for once it seems that consistency does not cost too much here.

As the initial goal of this PR is just the naming change, I suggest to follow-up with an issue/PR for providing per-language examples for those attributes in order to avoid any ambiguity in the future.

Reply:
I have opened an issue #1677 to keep track of the per-language definition and examples for each.